File: CSharp/Oomph.Data/UF09Test/UFs/v301/Int1/ATC001_B.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Oomph.Data.UF09Lib.UFs.v301;

namespace UF09Test.UFs.v301.Int1
{
	// Test: https://atcoder.jp/contests/atc001/tasks/unionfind_a
	// Test: https://atcoder.jp/contests/practice2/tasks/practice2_a
	class ATC001_B
	{
		static int[] Read() => Array.ConvertAll(Console.ReadLine().Split(), int.Parse);
		static (int, int) Read2() { var a = Read(); return (a[0], a[1]); }
		static void Main()
		{
			var (n, qc) = Read2();
			var qs = Array.ConvertAll(new bool[qc], _ => Read());
			var sb = new StringBuilder();

			var uf = new UnionFind(n);

			foreach (var q in qs)
			{
				if (q[0] == 0)
				{
					uf.Union(q[1], q[2]);
				}
				else
				{
					var b = uf.AreSame(q[1], q[2]);
					sb.AppendLine(b ? "Yes" : "No");
					//sb.Append(b ? 1 : 0).AppendLine();
				}
			}
			Console.Write(sb);
		}
	}
}

File: Python/oomph-data-ws/uf-309-lib/ATC001-B.py
# https://atcoder.jp/contests/atc001/tasks/unionfind_a

from unionfind306 import UnionFind

n, q = map(int, input().split())

uf = UnionFind(n)
r = []

for i in range(q):
    t, u, v = map(int, input().split())
    if t == 0:
        uf.union(u, v)
    else:
        r.append("Yes" if uf.are_same(u, v) else "No")

print("\n".join(r))
